google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITForeignKeyDeleteCascadeTest.java

```diff
@@ -0,0 +1,540 @@
+              transaction -> {
+                transaction.buffer(
+                    Arrays.asList(
+                        Mutation.newInsertBuilder("Singer")
+                            .set(getSingerIdColumnName()).to(4L)
+                            .set(getSingerFirstNameColumnName()).to("singerName").build(),
+                        Mutation.delete("Singer", Key.of(4L))));
+                return null;
+              });
+      fail("Expected exception");
+    } catch (SpannerException ex) {
+      assertThat(ex.getErrorCode()).isEqualTo(ErrorCode.FAILED_PRECONDITION);
+    }
+  }
+
+  @Test
+  public void testForeignKeyDeleteCascadeConstraints_forMutations_onConflictsDueToChildTable() {
+    final DatabaseClient databaseClient = getCreatedDatabaseClient();
+
+    // referencing a foreign key in child table and deleting the referenced key in parent table
+    // within the same mutations are considered conflicting operations.
+    try {
+      final String singerInsertStatement = getInsertStatementForSingerTable();
+      final Statement singerInsertStatementWithValues = Statement.newBuilder(singerInsertStatement)
+          // Use 'p1' to bind to the parameter with index 1 etc.
+          .bind("p1").to(5L)
+          .bind("p2").to("singerName").build();
+
+      databaseClient
+          .readWriteTransaction()
+          .run(
+              transaction -> {
+                transaction.executeUpdate(singerInsertStatementWithValues);
+                return null;
+              });
+
+      databaseClient
+          .readWriteTransaction()
+          .run(
+              transaction -> {
+                transaction.buffer(
+                    Arrays.asList(
+                        Mutation.newInsertBuilder("Concerts")
+                            .set(getConcertVenueIdColumnName()).to(5L)
+                            .set(getSingerIdColumnName()).to(5L).build(),
+                        Mutation.delete("Singer", Key.of(5L))));
+                return null;
+              });
+      fail("Expected exception");
+    } catch (SpannerException ex) {
+      assertThat(ex.getErrorCode()).isEqualTo(ErrorCode.FAILED_PRECONDITION);
+    }
+  }
+
+  private Database createDatabase(final List<String> statements) {
+    final Database database;
+    if (dialect.dialect == Dialect.POSTGRESQL) {
+      database = env.getTestHelper()
+          .createTestDatabase(Dialect.POSTGRESQL, statements);
+    } else {
+      database = env.getTestHelper()
+          .createTestDatabase(statements);
+    }
+    return database;
+  }
+
+  private String getReferentialConstraintsQueryStatement() {
+    if (dialect.dialect == Dialect.POSTGRESQL) {
+      return POSTGRES_QUERY_REFERENTIAL_CONSTRAINTS;
+    } else {
+      return QUERY_REFERENTIAL_CONSTRAINTS;
+    }
+  }
+
+  private String getReferentialConstraintsQueryStatementV2() {
+    if (dialect.dialect == Dialect.POSTGRESQL) {
+      return POSTGRES_QUERY_REFERENTIAL_CONSTRAINTS_V2;
```

Comment:
I had to create a different set of tables for one of the test cases since all other test cases operate on a common set of tables created in `@BeforeClass` method.

The V2 constants don't really have any different functionality. They just operate on the `V2` tables.

The only alternate I can think of is using a different set of entities for naming the tables apart from Singers/Concerts. In my opinion that could be an overkill. Scrolling to the constant value does make it clear that the V2 constants are used for the different tables.

Let me know what you think?